### PR TITLE
Add support for editing expenses

### PR DESCRIPTION
**Pull request by Hrishabh17**
## Summary by CodeRabbit

- **New Features**
  - Added functionality to update expenses when the action is 'UPDATED' and the resource is 'EXPENSE'.
  - Introduced request validation to prevent workspace mismatches.

- **Improvements**
  - Enhanced test fixtures with detailed expense-related information for more comprehensive testing.

- **Bug Fixes**
  - Fixed sorting of Line items in QuickBooks Online integration tests to ensure consistency.

- **Tests**
  - Added tests for the new `update_non_exported_expenses` functionality.
  - Included validation tests for request accuracy and workspace matching.

### CODE REVIEW

File: apps/fyle/tasks.py [hunk 1]
```diff
@@ -289,3 +290,28 @@ def import_and_export_expenses(report_id: str, org_id: str) -> None:
...
+
+    expense = Expense.objects.filter(org_id=org_id, expense_id=expense_id).first()
+
+    if expense:
+        if 'state' in expense.accounting_export_summary:
+            expense_state = expense.accounting_export_summary['state']
+        else:
+            expense_state = 'NOT_EXPORTED'
+
+    if expense_state and expense_state not in ['COMPLETE', 'IN_PROGRESS']:
```
Comment (ashwin1111) [inline apps/fyle/tasks.py:311]: move this 1 block inside

File: tests/test_fyle/test_tasks.py [hunk 2]
```diff
@@ -148,3 +149,35 @@ def test_sync_dimension(db, mocker):
     mock_platform_instance.import_fyle_dimensions.assert_called_once_with(is_export=True)
     mock_platform_instance.categories.sync.assert_called_once()
     mock_platform_instance.projects.sync.assert_called_once()
+
+
+def test_update_expenses(db, create_temp_workspace, mocker):
```
Comment (ashwin1111) [inline tests/test_fyle/test_tasks.py:154]: 1 test for exported expense case plis

File: apps/fyle/tasks.py [hunk 3]
```diff
@@ -292,26 +292,27 @@ def import_and_export_expenses(report_id: str, org_id: str) -> None:
...
+def update_non_exported_expenses(data: Dict) -> None:
     """
     To update expenses not in COMPLETE, IN_PROGRESS state
     """
     expense_state = None
     org_id = data['org_id']
     expense_id = data['id']
 
-    expense = Expense.objects.filter(org_id=org_id, expense_id=expense_id).first()
+    workspace_id = Workspace.objects.get(fyle_org_id = org_id)
```
Comment (ashwin1111) [inline apps/fyle/tasks.py:303]: let's add 1 assert and check if db ka workspace_id and URL path ka workspace_id is same, if not throw 4xx. This is an open endpoint, so having this check would be good

**Review by ashwin1111 — APPROVED** [review]

File: apps/fyle/tasks.py (hunk 3 above)
Comment (Hrishabh17) [reply apps/fyle/tasks.py]: added an assertion to check the url workspace_id and the workspace id with the org_id

File: apps/fyle/helpers.py [hunk 4]
```diff
@@ -260,6 +261,16 @@ def get_batched_expenses(batched_payload: List[dict], workspace_id: int) -> List
     return Expense.objects.filter(expense_id__in=expense_ids, workspace_id=workspace_id)
 
 
+def assert_valid_request(workspace_id:int, fyle_org_id:str):
```
Comment (ashwin1111) [inline apps/fyle/helpers.py:264]: workspace_id might be str, just re-confirm once. Convert to int if required

File: apps/fyle/helpers.py [hunk 5]
```diff
@@ -260,6 +261,16 @@ def get_batched_expenses(batched_payload: List[dict], workspace_id: int) -> List
...
+    """
+    Assert if the request is valid by checking
+    the url_workspace_id and fyle_org_id workspace
+    """
+    workspace = Workspace.objects.get(fyle_org_id=fyle_org_id)
+    if workspace.id != workspace_id:
+        raise ValidationError('Workspace mismatch')
```
Comment (ashwin1111) [inline apps/fyle/helpers.py:271]: logger.error() as well

File: apps/fyle/helpers.py (hunk 4 above)
Comment (Hrishabh17) [reply apps/fyle/helpers.py:264]: its int

File: apps/fyle/helpers.py (hunk 5 above)
Comment (Hrishabh17) [reply apps/fyle/helpers.py:271]: will add it

Comment (Hrishabh17) [reply apps/fyle/helpers.py:271]: we have logger.exception in place already.

**Pull request merged**